###Block 1

In [22]:
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
data = sns.load_dataset('geyser')

###Block 2

In [23]:
def k_fold_function(X,y,folds):
  training_score = []
  testing_score = []

  len_fold = (data.shape[0])//folds 

  #splitting data into train and test sets
  Xtrain = []
  Xtest = []

  ytrain = []
  ytest = []

  for i in range(folds):
    #length for the last fold as the dataset is uneven
    if i == (folds - 1):
      Xtrain = (X[:i*len_fold])
      Xtest = (X[i*len_fold:])
      ytrain = (y[:i*len_fold])
      ytest = (y[i*len_fold:])
    #length for the rest of the folds
    else:
      Xtrain = np.append(X[:i*len_fold], X[((i+1)*len_fold):])
      Xtest = (X[i*len_fold:len_fold*(i+1)])
      ytrain = np.append(y[:i*len_fold], y[((i+1)*len_fold):])
      ytest = (y[i*len_fold:len_fold*(i+1)])

    #print(Xtrain.shape) 
    #print(ytrain.shape) 
    Xtrain = Xtrain.reshape(Xtrain.shape[0],1)
    Xtest = Xtest.reshape(Xtest.shape[0],1)

    model = LinearRegression()
    model.fit(Xtrain,ytrain)
    training_score.append(model.score(Xtest, ytest))
    testing_score.append(model.score(Xtrain, ytrain))

  return training_score,testing_score

###Block 3

In [24]:
X = np.array(data['duration'])
y = np.array(data['waiting'])
training_score, testing_score = k_fold_function(X,y,3)
print("k fold = 3")
print("Average training score " + str(np.mean(training_score)))
print("Average testing score " + str(np.mean(testing_score)))
print("Std training score " + str(np.std(training_score)))
print("Std testing score : " + str(np.std(testing_score)))


training_score, testing_score = k_fold_function(X,y,5)
print("k fold = 5")
print("Average training score: " + str(np.mean(training_score)))
print("Average testing score: " + str(np.mean(testing_score)))
print("Std training score: " + str(np.std(training_score)))
print("Std testing score: " + str(np.std(testing_score)))

training_score, testing_score = k_fold_function(X,y,10)
print("k fold = 10")
print("Average training score: " + str(np.mean(training_score)))
print("Average testing score: " + str(np.mean(testing_score)))
print("Std training score: " + str(np.std(training_score)))
print("Std testing score: " + str(np.std(testing_score)))


k fold = 3
Average training score 0.8004576648847967
Average testing score 0.8121771711262316
Std training score 0.029830869135937025
Std testing score : 0.011968043511039845
k fold = 5
Average training score: 0.8056715142563547
Average testing score: 0.8116959665550336
Std training score: 0.020871940815441792
Std testing score: 0.004389581073677524
k fold = 10
Average training score: 0.7972280065767442
Average testing score: 0.8115509101080287
Std training score: 0.06683036198666513
Std testing score: 0.00522525331079697


###Block 4

Because the folds split evenly and the data are not shuffled, the training and testing R^2 values are fairly constant throughout the folds. The average testing score throughout different folds is around 0.80, with an average testing score of 0.81. Both the training and testing scores have a small difference, suggesting that the model chosen is a good model!